sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,742 @@
+            "Kafka record's timestamp is not 'LOG_APPEND_TIME' "
+                + "(topic: %s, partition %s, offset %s, timestamp type '%s')",
+            record.getTopic(),
+            record.getPartition(),
+            record.getOffset(),
+            record.getTimestampType());
+        return new Instant(record.getTimestamp());
+      };
+    }
+  }
+
+  /**
+   * A SplittableDoFn which reads from {@link Row} in {@link KafkaSourceDescriptionSchemas} which
+   * represents a Kafka source description and outputs {@link KafkaRecord}. By default, a {@link
+   * MonotonicallyIncreasing} watermark estimator is used to track watermark.
+   */
+  @VisibleForTesting
+  class ReadFromKafkaDoFn extends DoFn<Row, KafkaRecord<K, V>> {
+
+    ReadFromKafkaDoFn() {}
+
+    private final Map<String, Object> consumerConfig = ReadViaSDF.this.getConsumerConfig();
+
+    private final Map<String, Object> offsetConsumerConfig =
+        ReadViaSDF.this.getOffsetConsumerConfig();
+
+    private final DeserializerProvider keyDeserializerProvider =
+        ReadViaSDF.this.getKeyDeserializerProvider();
+    private final DeserializerProvider valueDeserializerProvider =
+        ReadViaSDF.this.getValueDeserializerProvider();
+
+    private final SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+        consumerFactoryFn = ReadViaSDF.this.getConsumerFactoryFn();
+    private final SerializableFunction<KafkaRecord<K, V>, Instant> extractOutputTimestampFn =
+        ReadViaSDF.this.getExtractOutputTimestampFn();
+
+    private final Duration kafkaPollTimeout = Duration.millis(1000);
+
+    // Variables that are initialized when bundle is started and closed when FinishBundle is called.
+    private transient ConsumerSpEL consumerSpEL = null;
+    private transient Deserializer<K> keyDeserializerInstance = null;
+    private transient Deserializer<V> valueDeserializerInstance = null;
+
+    private transient KafkaIOUtils.MovingAvg avgRecordSize = null;
+    private transient KafkaIOUtils.MovingAvg avgOffsetGap = null;
+
+    /**
+     * A {@link GrowableOffsetRangeTracker.RangeEndEstimator} which uses a Kafka {@link Consumer} to
+     * fetch backlog.
+     */
+    private class KafkaLatestOffsetEstimator
+        implements GrowableOffsetRangeTracker.RangeEndEstimator {
+      private final Consumer<byte[], byte[]> offsetConsumer;
+      private final TopicPartition topicPartition;
+      private final ConsumerSpEL consumerSpEL;
+
+      KafkaLatestOffsetEstimator(
+          Consumer<byte[], byte[]> offsetConsumer, TopicPartition topicPartition) {
+        this.offsetConsumer = offsetConsumer;
+        this.topicPartition = topicPartition;
+        this.consumerSpEL = new ConsumerSpEL();
+        this.consumerSpEL.evaluateAssign(
+            this.offsetConsumer, ImmutableList.of(this.topicPartition));
+      }
+
+      @Override
+      protected void finalize() {
+        try {
+          Closeables.close(offsetConsumer, true);
```

Comment:
Why would we close this when the object goes out of scope?

Would it make sense to have support for a KafkaConsumer pool that expires automatically after X amount of time of not being used?

Reply:
> Why would we close this when the object goes out of scope?

I need to close the connection when there is no more usage. There are 2 option:
1. The connection is only created when it's needed, like when querying the backlog.
2. The connection is created when the tracker is created and keep until the track is out of scope.
I go with 2 to avoid creating the same connection multiple time.

> Would it make sense to have support for a KafkaConsumer pool that expires automatically after X amount of time of not being used?

For Kafka, it's important to close consumer effectively to benefit from Kafka dynamic load balancing mechanism. But we are using manual assignment now. So as long as the connection can be created cheaply, we may not need the pool.

Reply:
Relying on finalize leads to having the JVM perform closing during garbage collection which means you can't control the timing of it and it could be a long time.

Lets stick with this for now and maybe there will be a better suggestion later if we think it will be a problem.